### ISSUE TITLE
Lint for use of single `&` and `|` in `bool` statements

### ISSUE THREAD

**Issue by alice-i-cecile** [rust-lang/rust-clippy#9619]
### What it does

Checks for the use of `&` and `|` in boolean statements where `&&` and `||` are preferred.

The latter are short-circuiting, while the former are bitwise operators.

Logical statements that combine bools (but not other types) should always prefer `&&` and `||`.

### Lint Name

nonlazy_boolean_operators

### Category

suspicious, perf

### Advantage

The `&&` and `||` lazy boolean operators short circuit, while `&` and `|` are intended for bitwise opereations and do not.

See https://www.reddit.com/r/rust/comments/xznz7v/does_rust_check_second_condition_if_first_is/ for more context.

### Drawbacks

Fixing this lint could break behavior if later arguments invoke side effects when evaluated. This is generally a *very* bad idea though, and should be carefully documented.

### Example

```rust
if a & b | c {
```

Could be written as:

```rust
if a && b || c {
```

**Comment by llogiq** [comment]
Sometimes using `&` or `|` in lieu of short-circuiting ops is done to avoid branches, which would hurt perf. The short-circuiting ops should only be preferred if a potentially costly operation (e.g. function call) comes after them.

**Comment by ChaiTRex** [comment]
As pointed out, there are valid reasons for choosing both styles, and I really doubt that the lint is going to be smart enough to tell those cases apart.

**Comment by alice-i-cecile** [comment]
Ah right, I'd forgotten about reducing branches as a concern.

I agree; it'll be too tricky to distinguish the right pattern here at the lint level.

**Comment by mqudsi** [comment]
It’s also common enough to want to report aggregate status (did *any* fail?) but still *need* to execute each individual function call regardless of past success/failure.
